### PR TITLE
📝 Upgrade docs

### LINKED ISSUE #83: ✨Type checking imports for all type hints
## Task

Only import type hints when `TYPE_CHECKING` is true.

Must be present:
```py
from __future__ import annotations

from typing import TYPE_CHECKING
```

## Example

```py
from __future__ import annotations

from typing import TYPE_CHECKING
from dataclasses import dataclass


if TYPE_CHECKING:
    from typing import List


@dataclass
class Foo:
    bar: List[int] = None
```

### LINKED ISSUE #86: 📝 Document code further
Further document docstrings &/or create documentation where none is present but is needed.

### PR DESCRIPTION
**Pull request by ooliver1**
### Changes

-   `adds`: More docstrings
-   `fixed`: closes #86 and closes #83
-   `improvements`: Better numpy docstrings with the RestructuredText also changed

 ### Check off the following

-   [x] I have tested my changes with the current requirements
-   [x] My Code follows the pep8 code style.

Linted and fixed with autopep8 and flake8, line endings are all LF.

https://pincer-upgraded-docs.rtfd.io

### CODE REVIEW

**Comment by Sigmanificient** [comment]
Holy!

File: pincer/utils/directory.py [hunk 1]
```diff
@@ -1,14 +1,28 @@
...
+    ----------
+    path:
+        Path to change to.
+
+    Yields
+    ------
+    Generator[:class:`str`, Any, None]
+    """
+    current_path = getcwd()
+    chdir(path)
```
Comment (Lunarmagpie) [inline pincer/utils/directory.py:25]: This and the `chdir` after it are supposed to be `os.chrdir`. Its been fixed on my pr but I think the maintainers should take another look at this to make sure that the code styling is right.

File: pincer/middleware/__init__.py [hunk 2]
```diff
@@ -1,16 +1,32 @@
...
+    convert_message, reply, interaction_response_handler, interaction_handler,
+    interaction_create_middleware
+)
+from .message_create import message_create_middleware
+from .message_delete import on_message_delete_middleware
+from .message_update import message_update_middleware
+from .payload import payload_middleware
+from .ready import on_ready_middleware
+from .voice_state_update import voice_state_update_middleware
+
```
[the remark this answers is not in the document]
Comment (ooliver1) [reply pincer/middleware/__init__.py:29]: i have an idea on how to refactor it

File: pincer/middleware/__init__.py [hunk 3]
```diff
@@ -46,3 +62,14 @@ def get_middleware() -> Dict[str, Coro]:
...
+__all__ = (
+    "channel_create_middleware", "error_middleware",
+    "guild_create_middleware", "convert_message", "reply",
+    "interaction_response_handler", "interaction_handler",
+    "interaction_create_middleware", "message_create_middleware",
+    "on_message_delete_middleware", "message_update_middleware",
+    "payload_middleware", "on_ready_middleware",
+    "voice_state_update_middleware"
+)
```
[the remark this answers is not in the document]
Comment (ooliver1) [reply pincer/middleware/__init__.py:75]: i have an idea on how to refactor it

File: pincer/objects/guild/channel.py [hunk 4]
```diff
@@ -328,15 +344,24 @@ async def edit(
         ...
 
     async def edit(self, **kwargs):
-        return await super().edit(**kwargs)
+        """Edit a text channel with the given keyword arguments.
 
-    @property
-    def mention(self) -> str:
```
[the remark this answers is not in the document]
Comment (ooliver1) [reply pincer/objects/guild/channel.py:334]: it subclasses channel, i dont remember moving it but yeah

File: pincer/exceptions.py [hunk 5]
```diff
@@ -297,8 +267,26 @@ class GatewayError(HTTPError):
     """Error code 502."""
 
 
+class ServerError(HTTPError):
+    """Error code 5xx."""
+
+
+class NotFoundError(HTTPError):
```
[the remark this answers is not in the document]
Comment (Lunarmagpie) [reply pincer/exceptions.py]: This is fixed now

File: pincer/utils/__init__.py [hunk 6]
```diff
@@ -3,19 +3,21 @@
...
 
+
 __all__ = (
-    "APINullable", "APIObject", "Choices", "Coro", "Descripted",
-    "HTTPMeta", "MISSING", "Snowflake", "Task", "TaskScheduler", "Timestamp",
+    "APINullable", "APIObject", "Choices", "Coro", "chdir", "HTTPMeta",
+    "MISSING", "Snowflake", "Task", "TaskScheduler", "Timestamp",
     "choice_value_types", "convert", "get_index", "get_signature_and_params",
-    "should_pass_cls", "should_pass_ctx"
+    "should_pass_cls", "should_pass_ctx", "MissingType", "get_params"
```
[the remark this answers is not in the document]
Comment (Lunarmagpie) [reply pincer/utils/__init__.py:22]: This is fixed now

File: pincer/objects/guild/guild.py [hunk 7]
```diff
@@ -441,7 +322,7 @@ class Guild(APIObject):
     approximate_presence_count: APINullable[int] = MISSING
     channels: APINullable[List[Channel]] = field(default_factory=list)
     # TODO: Add type when type is known
-    hub_type: APINullable[Any] = MISSING
+    hub_type: APINullable[...] = MISSING
```
[the remark this answers is not in the document]
Comment (ooliver1) [reply pincer/objects/guild/guild.py:325]: changed in next commit

**Review by trag1c — APPROVED** [review]

**Comment by Arthurdw** [comment]
_(**Current implementation is broken**)_

File: pincer/middleware/interaction_create.py [hunk 8]
```diff
@@ -102,26 +102,36 @@ async def interaction_handler(
...
+
+    Returns
+    -------
+    Tuple[:class:`str`, List[:class:`~pincer.objects.app.interactions.Interaction`]]
+        ``on_interaction_create`` and an ``Interaction``
+    """  # noqa: E501
     interaction: Interaction = Interaction.from_dict(
         construct_client_dict(self, payload.data)
     )
-    await interaction.build()
```
Comment (Arthurdw) [inline pincer/middleware/interaction_create.py:124]: Revert this removal.

File: pincer/objects/guild/guild.py [hunk 9]
```diff
@@ -518,18 +421,10 @@ async def modify_member(self, _id: int, **kwargs) -> GuildMember:
...
     @classmethod
     def from_dict(cls, data) -> Guild:
-        """
-        Instantiate a new guild from a dictionary.
-
-        Also handles it if the guild isn't available.
-
-        :raises UnavailableGuildError:
-            Exception gets raised when guild is unavailable.
-        """
```
Comment (Arthurdw) [inline pincer/objects/guild/guild.py:528]: Revert docs removal.

File: pincer/utils/__init__.py [hunk 10]
```diff
@@ -3,19 +3,21 @@
...
 from .tasks import Task, TaskScheduler
 from .timestamp import Timestamp
 from .types import (
-    APINullable, Coro, MISSING, Choices, choice_value_types, Descripted
+    APINullable, Coro, MISSING, MissingType, Choices, choice_value_types
 )
 
+
 __all__ = (
-    "APINullable", "APIObject", "Choices", "Coro", "Descripted",
```
Comment (Arthurdw) [inline pincer/utils/__init__.py:17]: Why did Descripted get removed?

**Review by Arthurdw — CHANGES_REQUESTED** [review]

File: pincer/middleware/interaction_create.py (hunk 8 above)
Comment (Lunarmagpie) [reply pincer/middleware/interaction_create.py]: Fixed

File: pincer/objects/guild/guild.py (hunk 9 above)
Comment (Lunarmagpie) [reply pincer/objects/guild/guild.py:528]: Rewrote the docs

File: pincer/utils/__init__.py (hunk 10 above)
Comment (Lunarmagpie) [reply pincer/utils/__init__.py:17]: Fixed

**Review by Arthurdw — APPROVED** [review]
~ Runs without exceptions for me

**Review by trag1c — APPROVED** [review]

File: pincer/objects/guild/channel.py [hunk 11]
```diff
@@ -6,28 +6,58 @@
...
-    from ..message import Message
-    from ..guild.overwrite import Overwrite
-    from ..guild.thread import ThreadMetadata
-    from ..guild.member import GuildMember
-    from ..user import User
-    from ... import Client
-    from ...utils import APINullable, Snowflake, Timestamp
+    from typing import Dict, List, Optional, Union
+
+    from ...client import Client
```
Comment (Lunarmagpie) [inline pincer/objects/guild/channel.py:21]: These imports are a mess

File: pincer/objects/guild/guild.py [hunk 12]
```diff
@@ -3,43 +3,47 @@
...
+from ...utils.types import MISSING
 from ...utils.api_object import APIObject
 from ...utils.conversion import construct_client_dict
-from ...utils.types import MISSING, APINullable
 
 if TYPE_CHECKING:
-    from ... import Client
+    from typing import Dict, List, Optional
+
+    from .role import Role
```
Comment (Lunarmagpie) [inline pincer/objects/guild/guild.py:20]: These imports are a mess

File: pincer/objects/guild/invite.py [hunk 13]
```diff
@@ -3,54 +3,54 @@
...
+    from typing import List, Optional
+
+    from .guild import Guild
+    from ..user.user import User
+    from ..guild.channel import Channel
+    from ...utils.types import APINullable
     from ..guild.member import GuildMember
-    from ...utils import APINullable, Timestamp
+    from ...utils.timestamp import Timestamp
+    from ..app.application import Application
```
Comment (Lunarmagpie) [inline pincer/objects/guild/invite.py:22]: `..app.application` should be above `...utils`

File: pincer/objects/message/component.py [hunk 14]
```diff
@@ -4,64 +4,56 @@
...
 if TYPE_CHECKING:
-    from ..app.select_menu import SelectOption
-    from ..message.button import ButtonStyle
+    from typing import List
+
     from ..message.emoji import Emoji
-    from ...utils import APINullable
+    from ...utils.types import APINullable
+    from ..message.button import ButtonStyle
+    from ..app.select_menu import SelectOption
```
Comment (Lunarmagpie) [inline pincer/objects/message/component.py:18]: import order should be changed

File: pincer/objects/message/context.py [hunk 15]
```diff
@@ -2,39 +2,38 @@
...
 
 if TYPE_CHECKING:
-    from ..app import ClientCommandStructure
+    from typing import Optional, Union
+
+    from ..user.user import User
     from ..guild.member import GuildMember
-    from ..user import User
     from ...utils.snowflake import Snowflake
+    from ..app.command import ClientCommandStructure
```
Comment (Lunarmagpie) [inline pincer/objects/message/context.py:14]: should be higher

**Review by Lunarmagpie — CHANGES_REQUESTED** [review]
Guess Ill go fix all the problems now :(

**Review by Lunarmagpie — APPROVED** [review]

**Review by trag1c — APPROVED** [review]

**Review by trag1c — APPROVED** [review]

**Review by Arthurdw — APPROVED** [review]

**Comment by Arthurdw** [comment]
Aight so I found a few things while quickly looking at the preview:
```
client.py:

Client.throttler missing : in defaults notation.
@event, load_cog docs rename botclient to myclient

commands.py:

update to descripted in command decorator example
document keep in remove_command & remove_commands
```

**Review by Sigmanificient — APPROVED** [review]

**Comment by Lunarmagpie** [comment]
> Aight so I found a few things while quickly looking at the preview:
> 
> ```
> client.py:
> 
> Client.throttler missing : in defaults notation.
> @event, load_cog docs rename botclient to myclient
> 
> commands.py:
> 
> update to descripted in command decorator example
> document keep in remove_command & remove_commands
> ```

I've made all of the changes.

**Review by Sigmanificient — APPROVED** [review]

**Review by Endercheif — APPROVED** [review]

**Review by Endercheif — APPROVED** [review]

File: pincer/__init__.py [hunk 16]
```diff
@@ -35,15 +36,16 @@ class VersionInfo(NamedTuple):
...
+            + (
+                f'-{self.release_level}{self.serial}'
+                * (self.release_level is not None)
+            )
         )
 
 
-__version__ = VersionInfo(0, 9, 3)
+version_info = VersionInfo(0, 9, 3, release_level="alpha")
+__version__ = "0.9.3a"
```
[the remark this answers is not in the document]
Comment (Arthurdw) [reply pincer/__init__.py:62]: It seems like your changed has been reverted :/

[the remark this answers is not in the document]
Comment (Arthurdw) [reply pincer/__init__.py:62]: It seems like your change has been reverted :/
Can you undo the revert?

**Review by Arthurdw — CHANGES_REQUESTED** [review]
We  are almost there! 💖

File: pincer/__init__.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (Lunarmagpie) [reply pincer/__init__.py]: I've made the change. :)

**Comment by Endercheif** [comment]
@ooliver1 originally, I recall you changing the import of `__version__` to RegEx. Why was that?

In the [latest commit](https://github.com/Pincer-org/Pincer/commit/4124eefebf082a0519f2bb151c7cc1a51c3d577b) by @Lunarmagpie it was "reverted" back. Do you believe we should keep the change or not?

**Comment by Lunarmagpie** [comment]
> @ooliver1 originally, I recall you changing the import of `__version__` to RegEx. Why was that?
> 
> In the [latest commit](https://github.com/Pincer-org/Pincer/commit/4124eefebf082a0519f2bb151c7cc1a51c3d577b) by @Lunarmagpie it was "reverted" back. Do you believe we should keep the change or not?

Yeah, it was kinda broken. I thought it would be better to import directly than rewrite the regex whenever that part of `__init__.py` changed.

**Comment by ooliver1** [comment]
I did try to import but it would say weird stuff on readthedocs about importing outside top level package, I could try again though

File: pincer/middleware/interaction_create.py [hunk 17]
```diff
@@ -97,22 +99,33 @@ async def interaction_handler(
 
     kwargs = {**defaults, **params}
 
-    await interaction_response_handler(self, command, context, interaction,
-                                       kwargs)
+    await interaction_response_handler(
+        self, command, context, interaction, kwargs)
```
Comment (Arthurdw) [inline pincer/middleware/interaction_create.py:103]: If your're putting it like this, put the `)` on a newline.

**Review by Arthurdw — CHANGES_REQUESTED** [review]

**Review by trag1c — APPROVED** [review]

**Review by Arthurdw — APPROVED** [review]

**Review by Lunarmagpie — APPROVED** [review]

**Comment by ooliver1** [comment]
:0 just looking at gh at school ayo

**Comment by Sigmanificient** [comment]
Seems like doc built wrongly xD

**Pull request merged**